File: gen_dic.rb

class Increment

  def initialize(len, limit, initlen=1)
    @len = len
    @limit = limit - 1
    @array = Array.new
    initlen.times {|x|
      @array[x] = 0
    }
    @array[0] = nil
  end

  def inc(n)
    #桁数がlimitを超えていたら
    return nil if n > @limit

    #今の桁を足す 
    if @array[n].nil? #|| @array[n].empty?
      @array[n] = 0
      return @array
    else
      @array[n] += 1 
    end
    #n桁目が@lenを超えていたら今の桁を0にして、次の桁をincrement
    if @array[n] >= @len
      @array[n] = 0
      a = inc(n+1)
      #桁数がlimitを超えていたらnilを返す
      return nil if a.nil?
    end
    return @array
  end
end


if __FILE__ == $0
  if ARGV.size < 2 
    puts "ruby #{__FILE__} <start_len> <end_len>"
    exit
  end
  start_len = ARGV[0].to_i
  end_len = ARGV[1].to_i
  @str = %Q{0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ!@#$%^&*()-_=+\|`~[]\{\};:'",<.>/?}
  size = @str.length
  s = Increment.new(size, end_len, start_len)

  while 1
    ary = s.inc(0)
    break if ary.nil?
    ary.reverse_each { |x|
      print @str[x]
    }
    print "\n"
  end
end

File: gen_dic.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

char *strings = " 0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ!@#$%^&*()-_=+|`~[]{};:'\",<.>/?";
int len, limit, length;
int *array;
int init_value = 0;

int *inc(int n){
  int *a;
  if(n > limit)
    return NULL;
  if(array[n] == init_value)
  {
    array[n] = 1;
    return array;
  }else{
    array[n] += 1;
  }

  if(array[n] >= length)
  {
    array[n] = 1;
    a = inc(n+1);
    if(a == NULL)
      return NULL;
  }
  return array;
}
void main(int argc, char *argv[]){
  if( argc < 2 ){
    printf("gen_dic <start_len> <end_len>\n");
    return;
  }
  int *result;

  len = atoi(argv[1]);
  limit = atoi(argv[2])-1;
  printf("start:%d, limit:%d\n", len, limit);
  length = (int)strlen(strings);
  printf("strlen: %d\n", length);

  // init
  printf("initialize...\n");
  array = (int *)malloc(limit);
  printf("array malloc!\n");
  for(int i=0; i<limit; i++)
    array[i] = init_value;

  //array[0] = init_value;
  printf("initialize complete!\n");

  //main
  while(1){
    result = inc(0);
    if(result == NULL)
      return;

    // print array.reverse_each
    for(int i=limit; i>=0; i--){
      if(array[i] != init_value)
        printf("%c", strings[array[i]]);
    }
    printf("\n");
  }
}

